### PR TITLE
feat: add Rocicorp Zero bindings

### PR DESCRIPTION
**Pull request by austinm911**
## Summary

Octane applications can now use Rocicorp Zero without a React compatibility layer. The new `@octanejs/zero` package ports Zero's public provider and hook surface while it keeps the database, sync, query, mutation, schema, and client implementations in `@rocicorp/zero`.

- Support `ZeroProvider`, `ZeroContext`, `useZero`, and `createUseZero`.
- Support reactive queries through `useQuery` and `useSuspenseQuery`.
- Support connection and online state through Octane's external-store hooks.
- Pin compatibility to `@rocicorp/zero@1.8.0` and record the canonical `zero/v1.8.0` source at commit `cdc02598f137ab4e071878f5674fdc716dbbc69d`.
- Vendor checksummed `zero-react` source and tests as review evidence. Exclude that evidence from the published package.
- Register the package in workspace metadata, generated binding catalogs, documentation, MCP discovery, and test infrastructure.

The port keeps the upstream materialized-view lifetime and subscription behavior. It replaces React context, nodes, Suspense integration, and hook calls with Octane equivalents. Conformance tests cover the public exports, provider ownership and option construction, auth lifecycle, subscriptions, query materialization, Suspense, navigation remounts, and view cleanup.

## Validation

- [x] `pnpm install --frozen-lockfile`
- [x] `pnpm sync`
- [x] `pnpm format:check`
- [x] `pnpm typecheck`
- [ ] `pnpm test` — The local full suite is blocked by unrelated integration-harness failures tracked in #568. With the proposed harness fixes applied separately, 1,537 test files and all 16,284 tests passed. Those harness fixes are intentionally excluded from this PR.
- [x] targeted tests:
  - Zero conformance tests: 15 passed.
  - MCP bridge tests: 23 passed.
  - Hook-slot registry tests: 60 passed.
  - Generated binding status, parity gap, package inventory, React parity, and test-marker checks passed.
  - Vendored upstream SHA-256 verification passed for all 20 files.
  - The package tarball contains only the license, package metadata, README, and intended source files.

## Risk / follow-ups

- This release supports exactly `@rocicorp/zero@1.8.0`.
- Zero query data does not load during server rendering. This matches the upstream binding.
- The public runtime and type surface is covered. Some upstream result-state, error, TTL, state-loss rotation, and auth-variant test cases remain explicit coverage gaps in `packages/zero/UPSTREAM.md`.

## Provenance

- [x] An agent produced this diff (`agent-authored`)

---

> [!NOTE]
> **Medium Risk**
> New reactive data/sync binding with Suspense and client lifecycle semantics; SSR does not load query data (documented parity with upstream). Some upstream edge-case tests are not yet adapted.
> 
> **Overview**
> Introduces **`@octanejs/zero`** (`0.0.1`), a new framework binding that ports the public **`@rocicorp/zero/react@1.8.0`** surface onto Octane while keeping sync, schema, and client logic in **`@rocicorp/zero`**.
> 
> The package adds **`ZeroProvider`**, context hooks (**`useZero`**, **`createUseZero`**), reactive **`useQuery`** / **`useSuspenseQuery`** (with upstream **`ViewStore`** materialization), and **`useConnectionState`** / **`useZeroOnline`**, wired through Octane **`useSyncExternalStore`** and **`use()`** for Suspense. **`ZeroProvider`** is implemented as **`.tsrx`** with **`OctaneNode`** children; option-built clients still instantiate only in a client **`useEffect`**, matching upstream SSR behavior.
> 
> Workspace wiring updates generated binding docs and inventories (54 bindings), CLI **`octane-data.json`** (including **`@rocicorp/zero/react` → `@octanejs/zero`**), MCP bridge discovery, hook-slot registry, and eval manifest lockfile hashes. Conformance tests cover exports parity, provider lifecycle, subscriptions, and query/Suspense paths; **`UPSTREAM.md`** documents vendored oracle source and remaining upstream test gaps.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ae6acf8538aef00a68a7169325ed2344575ca5c0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/zero/src/use-query.ts [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+	const q = query ? addContextToQuery(query, zero.context) : undefined;
+	const view = q ? viewStore.getView(zero, q, enabled, ttl) : undefined;
+
+	// https://react.dev/reference/react/useSyncExternalStore
+	// Always call useSyncExternalStore to maintain consistent hook order
+	return useSyncExternalStore(
+		view?.subscribeReactInternals ?? disabledSubscriber,
+		view?.getSnapshot ?? (getDisabledSnapshot as () => MaybeQueryResult<TReturn>),
+		view?.getSnapshot ?? (getDisabledSnapshot as () => MaybeQueryResult<TReturn>),
+	);
```
[the remark this answers is not in the document]
Comment (austinm911) [reply packages/zero/src/use-query.ts:133]: > `useQuery` and `useSuspenseQuery` take an optional trailing `options` argument, but the package declares manual hook slots and never strips the compiler-injected trailing Symbol.

Fixed in 864268d8. Both hooks now separate the compiler slot from user options and forward it to `useSyncExternalStore`. A compiled `.tsrx` regression also verifies that a one-argument Suspense query keeps Zero's default `partial` threshold.

File: packages/zero/src/use-query.ts [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+
+	if (view && enabled) {
+		if (suspendUntil === 'complete' && !view.complete) {
+			use(view.waitForComplete());
+		}
+
+		if (suspendUntil === 'partial' && !view.nonEmpty) {
+			use(view.waitForNonEmpty());
+		}
+	}
```
[the remark this answers is not in the document]
Comment (austinm911) [reply packages/zero/src/use-query.ts:261]: > `useSuspenseQuery` only calls `use()` while the view is still waiting, then skips it once ready.

Fixed in af70fea3. Each query call site now retains its settled promise in the dynamic `use()` sequence and uses distinct sub-slots for its store and promise state. A compiled two-query regression verifies that the boundary stays pending after the first result completes and reveals only after the second result completes.

File: packages/zero/src/use-query.ts [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+			cache.promise.status !== 'pending')
+	) {
+		cache.view = view;
+		cache.suspendUntil = suspendUntil;
+		cache.promise = (
+			suspendUntil === 'complete' ? view.waitForComplete() : view.waitForNonEmpty()
+		) as SuspensePromise;
+	}
+
+	if (cache.promise !== undefined) use(cache.promise);
```
[the remark this answers is not in the document]
Comment (austinm911) [reply packages/zero/src/use-query.ts]: > `useSuspensePromise` keeps calling `use` on a cached pending promise after `shouldSuspend` becomes false.

Good catch. This is the cancellation side of retaining a stable dynamic `use()` position. I am keeping this thread open while choosing whether the retained promise should be locally cancelable or whether `ViewWrapper` should own invalidation of its wait promises.

[the remark this answers is not in the document]
Comment (austinm911) [reply packages/zero/src/use-query.ts]: > `useSuspensePromise` keeps calling `use` on a cached pending promise after `shouldSuspend` becomes false.

Fixed in 187fb171. The hook now races each Zero wait promise with a local cancellation promise. Disabling, clearing, or replacing the query settles the retained dynamic `use()` position without changing `ViewWrapper` ownership or materialization behavior.

New compiled regressions disable and clear an unresolved query and verify that the pending boundary releases. The existing sequential-query regression still verifies stable call order. All 14 Zero tests pass, and the full `pnpm test` run passes all 16,283 tests.

File: packages/zero/src/use-query.ts [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+		let cancel = () => {};
+		const cancellation = new Promise<void>((resolve) => {
+			cancel = resolve;
+		});
+		const source = suspendUntil === 'complete' ? view.waitForComplete() : view.waitForNonEmpty();
+
+		cache.cancel = cancel;
+		cache.view = view;
+		cache.suspendUntil = suspendUntil;
+		cache.promise = Promise.race([source, cancellation]) as SuspensePromise;
```
[the remark this answers is not in the document]
Comment (austinm911) [reply packages/zero/src/use-query.ts]: > If `shouldSuspend` is true again when that cancellation settles, resume replay creates a fresh wait promise but Octane reuses the cancelled fulfillment.

Fixed in 37b9d10f by removing cancellation from this path. The hook now reserves its stable dynamic `use()` position with a shared fulfilled sentinel whenever the query does not suspend. It uses the current Zero wait promise only while the query must suspend.

A compiled regression disables the unresolved query and enables it in the next microtask. It verifies that the boundary suspends again until Zero reports a complete result. All 15 Zero tests pass, and the full `pnpm test` run passes all 16,284 tests.

**Comment by leonidaz** [comment]
This React library port does not satisfy the Octane criterion of 1 million npm downloads per week. Across the official Rocicorp Zero package family, the most-downloaded package is @rocicorp/zero, with 103,341 downloads during 2026-08-02 through 2026-08-08.

If you strongly feel that this library needs to be ported, or if the original library author is willing to take on maintenance of the package after it is ported, please create an issue and explain your reasons.

We thank you for your efforts creating this PR, but at this point we cannot take on the burden of maintaining all of these packages.
